File: src/poker/bot_runner.lua
--- Bot runner: joins a table and loops, calling a strategy function each turn.
---
--- Usage:
---   local bot_runner = require("poker.bot_runner")
---   bot_runner.run({
---     strategy = function(state, me) return { action = "call" } end,
---     url  = "http://127.0.0.1:8080",
---     table_id  = "demo",
---     player_id = "LuaBot",
---     chips     = 500,
---   })

local Client = require("poker.client")

local M = {}

local function build_me(state, player_id)
  local seats = state.seats or {}
  local hand = state.hand or {}
  local hc_map = hand.hole_cards or {}
  local contrib_map = hand.contribution or {}
  local seat, stack = nil, 0

  for i, s in ipairs(seats) do
    if type(s) == "table" and s.player_id == player_id then
      seat = i
      stack = s.stack or 0
      break
    end
  end

  local hole_cards = {}
  if seat and type(hc_map) == "table" then
    hole_cards = hc_map[tostring(seat)] or {}
  end

  local contribution = 0
  if seat and type(contrib_map) == "table" then
    contribution = contrib_map[tostring(seat)] or 0
  end

  return {
    player_id = player_id,
    seat = seat,
    stack = stack,
    hole_cards = hole_cards,
    contribution = contribution,
  }
end

local function default_strategy(state, me)
  local hand = state.hand or {}
  local cb = hand.current_bet or 0
  local mri = hand.min_raise_increment or 5
  local stack = me.stack or 0
  local contrib = me.contribution or 0

  local target = cb + mri
  local need = target - contrib
  if need > 0 and need <= stack then
    return { action = "raise", amount = target }
  end

  local call_need = cb - contrib
  if call_need > 0 and call_need <= stack then
    return { action = "call" }
  end

  if contrib >= cb then
    return { action = "check" }
  end

  return { action = "fold" }
end

--- @param opts { strategy?: function, url?: string, table_id?: string, player_id?: string, chips?: number, max_hands?: number, verbose?: boolean }
function M.run(opts)
  opts = opts or {}
  local strategy = opts.strategy or default_strategy
  local url = opts.url or "http://127.0.0.1:8080"
  local table_id = opts.table_id or "demo"
  local player_id = opts.player_id or "LuaBot"
  local chips = opts.chips or 500
  local max_hands = opts.max_hands
  local verbose = opts.verbose ~= false

  local client, cerr = Client.new({ base_url = url })
  if not client then
    io.stderr:write("[bot] Client error: " .. tostring(cerr) .. "\n")
    return
  end

  if verbose then
    io.write("[bot] Connecting to " .. url .. ", table=" .. table_id .. ", name=" .. player_id .. "\n")
  end

  local resp, jerr = client:join_table(table_id, { player_id = player_id, chips = chips })
  if not resp then
    io.stderr:write("[bot] Join failed: " .. tostring(jerr and jerr.message or jerr) .. "\n")
    return
  end

  local state = resp.table or {}
  local me = build_me(state, player_id)
  if verbose then
    io.write("[bot] Seated at seat " .. tostring(me.seat) .. " with " .. chips .. " chips\n")
  end

  local hands_played = 0
  while max_hands == nil or hands_played < max_hands do
    if verbose then
      io.write("[bot] Waiting for turn…\n")
    end

    state = client:wait_for_turn(table_id, player_id, { timeout = 300 })
    if not state then
      if verbose then
        io.write("[bot] Timed out waiting. Retrying…\n")
      end
    else
      me = build_me(state, player_id)
      local hand = state.hand or {}

      if verbose then
        local cards = table.concat(me.hole_cards or {}, " ")
        if cards == "" then cards = "??" end
        local comm_list = hand.community or {}
        local comm = table.concat(comm_list, " ")
        io.write("[bot] Hand: [" .. cards .. "]  Community: [" .. comm .. "]  "
          .. "Pot: " .. tostring(hand.pot or 0) .. "  Bet: " .. tostring(hand.current_bet or 0)
          .. "  Stack: " .. tostring(me.stack) .. "\n")
      end

      local ok_strat, decision = pcall(strategy, state, me)
      if not ok_strat then
        io.stderr:write("[bot] Strategy error: " .. tostring(decision) .. "\n")
        decision = { action = "fold" }
      end

      local action = decision.action or "fold"
      local amount = decision.amount

      local aresp, aerr = client:send_action(table_id, {
        player_id = player_id,
        action = action,
        amount = amount,
      })

      if verbose then
        if aresp then
          local amt_str = amount and (" " .. tostring(amount)) or ""
          io.write("[bot] -> " .. action .. amt_str .. "\n")
        else
          io.write("[bot] Action rejected: " .. tostring(aerr and aerr.message or aerr) .. "\n")
        end
      end

      local new_hand = aresp and aresp.table and aresp.table.hand or {}
      if new_hand.status == "idle" then
        hands_played = hands_played + 1
        if verbose then
          io.write("[bot] Hand finished (total: " .. hands_played .. ")\n")
        end
      end
    end
  end

  if verbose then
    io.write("[bot] Leaving table…\n")
  end
  pcall(function() client:leave_table(table_id, player_id) end)
end

return M

File: clients/python/bot_runner.py
"""
Bot runner: automatically plays poker using a user-supplied strategy function.

Usage::

    from bot_runner import run_bot

    def my_strategy(state, me):
        \"\"\"
        state  – full table state dict from GET /v1/tables/:id/state
        me     – dict with { player_id, seat, stack, hole_cards, contribution }

        Return a dict: { "action": "fold"|"check"|"call"|"raise"|"all_in",
                         "amount": <int> }   (amount only needed for raise/bet)
        \"\"\"
        hand = state["hand"]
        cb = hand["current_bet"]
        mri = hand["min_raise_increment"]
        return {"action": "raise", "amount": cb + mri}

    run_bot(my_strategy)

CLI::

    python bot_runner.py                     # uses built-in min-raise bot
    python bot_runner.py my_bot.py           # loads decide(state, me) from file
    python bot_runner.py --name Bot1 --chips 1000 --url http://localhost:8080
"""

from __future__ import annotations

import argparse
import importlib.util
import json
import os
import sys
import time
import traceback
from typing import Any, Callable, Optional

_ROOT = os.path.dirname(os.path.abspath(__file__))
if _ROOT not in sys.path:
    sys.path.insert(0, _ROOT)

from poker_client import PokerClient, PokerError, TransportError

StrategyFn = Callable[[dict, dict], dict]


def _build_me(state: dict, player_id: str) -> dict:
    """Extract convenience info about our player from the state."""
    seats = state.get("seats") or []
    hand = state.get("hand") or {}
    hole_cards_map = hand.get("hole_cards") or {}
    contrib_map = hand.get("contribution") or {}

    seat = None
    stack = 0
    for i, s in enumerate(seats, 1):
        if isinstance(s, dict) and s.get("player_id") == player_id:
            seat = i
            stack = s.get("stack", 0)
            break

    hole_cards = []
    if isinstance(hole_cards_map, dict) and seat is not None:
        hole_cards = hole_cards_map.get(str(seat), [])

    contribution = 0
    if isinstance(contrib_map, dict) and seat is not None:
        contribution = contrib_map.get(str(seat), 0)

    return {
        "player_id": player_id,
        "seat": seat,
        "stack": stack,
        "hole_cards": hole_cards,
        "contribution": contribution,
    }


def _default_strategy(state: dict, me: dict) -> dict:
    """Built-in fallback: min-raise when possible, else call/all-in, else check."""
    hand = state.get("hand") or {}
    cb = hand.get("current_bet", 0)
    mri = hand.get("min_raise_increment", 5)
    stack = me.get("stack", 0)
    contrib = me.get("contribution", 0)
    target = cb + mri
    need = target - contrib
    if need > 0 and need <= stack:
        return {"action": "raise", "amount": target}
    call_need = cb - contrib
    if call_need > 0 and call_need <= stack:
        return {"action": "call"}
    if call_need > 0 and stack > 0:
        return {"action": "all_in"}
    if contrib >= cb:
        return {"action": "check"}
    return {"action": "fold"}


def _load_strategy_file(path: str) -> StrategyFn:
    """Import a .py file and return its `decide(state, me)` function."""
    spec = importlib.util.spec_from_file_location("user_bot", path)
    if spec is None or spec.loader is None:
        raise RuntimeError(f"Cannot load module from {path}")
    mod = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(mod)
    fn = getattr(mod, "decide", None)
    if fn is None:
        raise RuntimeError(
            f"{path} must define a top-level function: decide(state, me) -> dict"
        )
    return fn


def run_bot(
    strategy: Optional[StrategyFn] = None,
    *,
    url: str = "http://127.0.0.1:8080",
    table_id: str = "demo",
    player_id: str = "PythonBot",
    chips: Optional[int] = None,
    max_hands: Optional[int] = None,
    poll_interval: float = 0.5,
    verbose: bool = True,
) -> None:
    """
    Join a table and loop forever (or for *max_hands*), calling *strategy*
    each time it is our turn.
    """
    if strategy is None:
        strategy = _default_strategy

    client = PokerClient(url)

    if verbose:
        print(f"[bot] Connecting to {url}, table={table_id}, name={player_id}")

    try:
        resp = client.join_table(table_id, player_id=player_id, chips=chips)
    except PokerError as e:
        print(f"[bot] Join failed: {e}", file=sys.stderr)
        raise SystemExit(1)

    state = resp.get("table", {})
    me = _build_me(state, player_id)
    if verbose:
        print(f"[bot] Seated at seat {me['seat']} with {me.get('stack', '?')} chips")

    hands_played = 0
    try:
        while max_hands is None or hands_played < max_hands:
            if verbose:
                print(f"[bot] Waiting for turn…")

            try:
                state = client.wait_for_turn(
                    table_id, player_id, poll_interval=poll_interval, timeout=300,
                )
            except TimeoutError:
                if verbose:
                    print("[bot] Timed out waiting (5 min). Retrying…")
                continue

            me = _build_me(state, player_id)
            hand = state.get("hand") or {}

            if verbose:
                cards = " ".join(me["hole_cards"]) if me["hole_cards"] else "??"
                comm = " ".join(str(c) for c in (hand.get("community") or []))
                print(
                    f"[bot] Hand: [{cards}]  Community: [{comm}]  "
                    f"Pot: {hand.get('pot', 0)}  Bet: {hand.get('current_bet', 0)}  "
                    f"Stack: {me['stack']}"
                )

            try:
                decision = strategy(state, me)
            except Exception:
                traceback.print_exc()
                decision = {"action": "fold"}
                if verbose:
                    print("[bot] Strategy error — folding")

            action = decision.get("action", "fold")
            amount = decision.get("amount")

            try:
                resp = client.send_action(
                    table_id,
                    player_id=player_id,
                    action=action,
                    amount=amount,
                )
                if verbose:
                    amt_str = f" {amount}" if amount is not None else ""
                    print(f"[bot] -> {action}{amt_str}")
            except PokerError as e:
                if verbose:
                    print(f"[bot] Action rejected: {e.api_code} — {e.message}")

            new_hand = (resp or {}).get("table", {}).get("hand", {})
            if new_hand.get("status") == "idle":
                hands_played += 1
                if verbose:
                    print(f"[bot] Hand finished (total: {hands_played})")

    except KeyboardInterrupt:
        print("\n[bot] Interrupted")
    finally:
        if verbose:
            print(f"[bot] Leaving table…")
        try:
            client.leave_table(table_id, player_id=player_id)
        except Exception:
            pass


def main() -> None:
    default_url = os.environ.get("POKER_URL", "http://127.0.0.1:8080")
    p = argparse.ArgumentParser(description="Poker bot runner (Python)")
    p.add_argument(
        "bot_file", nargs="?", default=None,
        help="Path to a .py file with decide(state, me) -> dict",
    )
    p.add_argument("--url", default=default_url, help="Server URL")
    p.add_argument("--table", default="demo", help="Table ID")
    p.add_argument("--name", default="PythonBot", help="Player name")
    p.add_argument(
        "--chips",
        type=int,
        default=None,
        help="Deprecated: ignored by server; table buy_in_chips is used.",
    )
    p.add_argument("--hands", type=int, default=None, help="Max hands to play")
    p.add_argument("-q", "--quiet", action="store_true", help="Less output")
    args = p.parse_args()

    strategy = None
    if args.bot_file:
        strategy = _load_strategy_file(args.bot_file)
        print(f"[bot] Loaded strategy from {args.bot_file}")

    run_bot(
        strategy,
        url=args.url,
        table_id=args.table,
        player_id=args.name,
        chips=args.chips,  # optional; server uses table buy-in
        max_hands=args.hands,
        verbose=not args.quiet,
    )


if __name__ == "__main__":
    main()
